src/ott/neural/flows/otfm.py

```diff
@@ -0,0 +1,371 @@
+    )
+    ResampleMixin.__init__(self)
+    UnbalancednessMixin.__init__(
+        self,
+        rng=rng_unbalanced,
+        source_dim=input_dim,
+        target_dim=input_dim,
+        cond_dim=cond_dim,
+        tau_a=tau_a,
+        tau_b=tau_b,
+        rescaling_a=rescaling_a,
+        rescaling_b=rescaling_b,
+        unbalanced_kwargs=unbalanced_kwargs,
+    )
+
+    self.velocity_field = velocity_field
+    self.input_dim = input_dim
+    self.ot_solver = ot_solver
+    self.flow = flow
+    self.time_sampler = time_sampler
+    self.optimizer = optimizer
+    self.epsilon = epsilon
+    self.cost_fn = cost_fn
+    self.scale_cost = scale_cost
+    self.callback_fn = callback_fn
+    self.checkpoint_manager = checkpoint_manager
+    self.rng = rng
+    self.logging_freq = logging_freq
+    self.num_eval_samples = num_eval_samples
+    self._training_logs: Mapping[str, Any] = defaultdict(list)
+
+    self.setup()
+
+  def setup(self):
+    """Setup :class:`OTFlowMatching`."""
+    self.state_velocity_field = (
+        self.velocity_field.create_train_state(
+            self.rng, self.optimizer, self.input_dim
+        )
+    )
+
+    self.step_fn = self._get_step_fn()
+    if self.ot_solver is not None:
+      self.match_fn = self._get_sinkhorn_match_fn(
+          self.ot_solver,
+          epsilon=self.epsilon,
+          cost_fn=self.cost_fn,
+          scale_cost=self.scale_cost,
+          tau_a=self.tau_a,
+          tau_b=self.tau_b,
+      )
+    else:
+      self.match_fn = None
+
+  def _get_step_fn(self) -> Callable:
+
+    @jax.jit
+    def step_fn(
+        key: jax.random.PRNGKeyArray,
+        state_velocity_field: train_state.TrainState,
+        batch: Dict[str, jnp.ndarray],
+    ) -> Tuple[Any, Any]:
+
+      def loss_fn(
+          params: jnp.ndarray, t: jnp.ndarray, noise: jnp.ndarray,
+          batch: Dict[str, jnp.ndarray], rng: jax.random.PRNGKeyArray
+      ) -> jnp.ndarray:
+
+        x_t = self.flow.compute_xt(
+            noise, t, batch["source_lin"], batch["target_lin"]
+        )
+        apply_fn = functools.partial(
+            state_velocity_field.apply_fn, {"params": params}
+        )
+        v_t = jax.vmap(apply_fn)(
+            t=t, x=x_t, condition=batch["source_conditions"], rng=rng
+        )
+        u_t = self.flow.compute_ut(t, batch["source_lin"], batch["target_lin"])
+        return jnp.mean((v_t - u_t) ** 2)
+
+      batch_size = len(batch["source_lin"])
+      key_noise, key_t, key_model = jax.random.split(key, 3)
+      keys_model = jax.random.split(key_model, batch_size)
+      t = self.time_sampler(key_t, batch_size)
+      noise = self.sample_noise(key_noise, batch_size)
+      grad_fn = jax.value_and_grad(loss_fn)
+      loss, grads = grad_fn(
+          state_velocity_field.params, t, noise, batch, keys_model
+      )
+      return state_velocity_field.apply_gradients(grads=grads), loss
+
+    return step_fn
+
+  def __call__(self, train_loader, valid_loader):
```

Comment:
I'm thinking whether it's necessary to have such a think wrapper over a training loop - maybe it would be better if we had sort of a step function for the `OTFM` class (as well as `GENOT`, etc.) that takes in the batch and does just one step + returns some metadata (such as loss, etc.). This way, we don't need to pass many arguments (`iterations`, `valid_freq` and `checkpoint_manager`) and leave this for the user, which gives more flexibility and makes the code a bit more maintainable.
Wdyt?

Reply:
Ideally if you agree, we'd have to do this for all the other models as well (like `W2NeuralDual`, which would require more refactoring... The benefit of this approach is that we'd have a trainer class which could accept any model (`W2NeuralDual`, `OTFM`, `GENOT`, `MapEstimator` for Monge gap) + it would align itself more with the composition over inheritance type of approach.

Reply:
Let's discuss this in person!